// @eagleoutice/flowr-dev/cli/repl/commands/repl-version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionCommand = void 0;
exports.retrieveVersionInformation = retrieveVersionInformation;
exports.printVersionInformation = printVersionInformation;
const version_1 = require("../../../util/version");
const assert_1 = require("../../../util/assert");
const versionRegex = /^\d+\.\d+\.\d+/m;
async function retrieveVersionInformation(shell) {
    const flowr = (0, version_1.flowrVersion)().toString();
    const r = (await shell.usedRVersion())?.format() ?? 'unknown';
    (0, assert_1.guard)(versionRegex.test(flowr), `flowR version ${flowr} does not match the expected format!`);
    (0, assert_1.guard)(r === 'unknown' || versionRegex.test(r), `R version ${r} does not match the expected format!`);
    return { flowr: flowr, r: r };
}
async function printVersionInformation(output, shell) {
    const { flowr, r } = await retrieveVersionInformation(shell);
    output.stdout(`flowR: ${flowr}`);
    output.stdout(`R: ${r}`);
}
exports.versionCommand = {
    description: 'Prints the version of flowR as well as the current version of R',
    aliases: [],
    usageExample: ':version',
    script: false,
    fn: (output, shell) => printVersionInformation(output, shell)
};
