// @eagleoutice/flowr-dev/cli/repl/server/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateBaseMessageFormat = validateBaseMessageFormat;
exports.validateMessage = validateMessage;
exports.answerForValidationError = answerForValidationError;
const send_1 = require("./send");
const messages_1 = require("./messages/messages");
function validateBaseMessageFormat(input) {
    try {
        return validateMessage(JSON.parse(input), messages_1.baseMessage);
    }
    catch (e) {
        return { type: 'error', reason: e };
    }
}
function validateMessage(input, def) {
    try {
        const result = def.schema.validate(input);
        return result.error ? { type: 'error', reason: result.error } : { type: 'success', message: input };
    }
    catch (e) {
        return { type: 'error', reason: e };
    }
}
function answerForValidationError(client, result, id) {
    (0, send_1.sendMessage)(client, {
        type: 'error',
        fatal: false,
        id: id,
        reason: `Invalid message format: ${result.reason.message}`
    });
}
