// @eagleoutice/flowr-dev/core/steps/all/core/00-parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PARSE_WITH_R_SHELL_STEP = void 0;
const print_1 = require("../../../print/print");
const parse_printer_1 = require("../../../print/parse-printer");
const retriever_1 = require("../../../../r-bridge/retriever");
function processor(_results, input) {
    /* in the future, we want to expose all cases */
    if (Array.isArray(input.request)) {
        return (0, retriever_1.retrieveParseDataFromRCode)(input.request[0], input.shell);
    }
    else {
        return (0, retriever_1.retrieveParseDataFromRCode)(input.request, input.shell);
    }
}
exports.PARSE_WITH_R_SHELL_STEP = {
    name: 'parse',
    humanReadableName: 'parse with R shell',
    description: 'Parse the given R code into an AST',
    processor,
    executed: 0 /* PipelineStepStage.OncePerFile */,
    printer: {
        [0 /* StepOutputFormat.Internal */]: print_1.internalPrinter,
        [2 /* StepOutputFormat.Json */]: text => text,
        [5 /* StepOutputFormat.RdfQuads */]: parse_printer_1.parseToQuads
    },
    dependencies: [],
    requiredInput: undefined
};
