// @eagleoutice/flowr-dev/core/steps/all/core/20-dataflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATIC_DATAFLOW = void 0;
const print_1 = require("../../../print/print");
const dataflow_printer_1 = require("../../../print/dataflow-printer");
const extractor_1 = require("../../../../dataflow/extractor");
const staticDataflowCommon = {
    name: 'dataflow',
    description: 'Construct the dataflow graph',
    executed: 0 /* PipelineStepStage.OncePerFile */,
    printer: {
        [0 /* StepOutputFormat.Internal */]: print_1.internalPrinter,
        [2 /* StepOutputFormat.Json */]: dataflow_printer_1.dataflowGraphToJson,
        [5 /* StepOutputFormat.RdfQuads */]: dataflow_printer_1.dataflowGraphToQuads,
        [3 /* StepOutputFormat.Mermaid */]: dataflow_printer_1.dataflowGraphToMermaid,
        [4 /* StepOutputFormat.MermaidUrl */]: dataflow_printer_1.dataflowGraphToMermaidUrl
    },
    dependencies: ['normalize'],
};
function legacyProcessor(results, input) {
    return (0, extractor_1.produceDataFlowGraph)(input.request, results.normalize);
}
exports.STATIC_DATAFLOW = {
    ...staticDataflowCommon,
    humanReadableName: 'dataflow',
    processor: legacyProcessor,
    requiredInput: {}
};
