// @eagleoutice/flowr-dev/statistics/features/feature.js
"use strict";
/**
 * This module holds the definition of what a {@link Feature} that can be extracted from an R AST is.
 *
 * Furthermore, it contains the definition of all features that are known in {@link ALL_FEATURES}.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.allFeatureNames = exports.ALL_FEATURES = void 0;
const used_packages_1 = require("./supported/used-packages/used-packages");
const comments_1 = require("./supported/comments/comments");
const defined_functions_1 = require("./supported/defined-functions/defined-functions");
const used_functions_1 = require("./supported/used-functions/used-functions");
const values_1 = require("./supported/values/values");
const assignments_1 = require("./supported/assignments/assignments");
const loops_1 = require("./supported/loops/loops");
const control_flow_1 = require("./supported/control-flow/control-flow");
const data_access_1 = require("./supported/data-access/data-access");
const statistics_expression_list_1 = require("./supported/expression-list/statistics-expression-list");
const variables_1 = require("./supported/variables/variables");
/**
 * The source of truth for all features that are supported by the statistics.
 */
exports.ALL_FEATURES = {
    usedPackages: used_packages_1.usedPackages,
    comments: comments_1.comments,
    definedFunctions: defined_functions_1.definedFunctions,
    usedFunctions: used_functions_1.usedFunctions,
    values: values_1.values,
    assignments: assignments_1.assignments,
    loops: loops_1.loops,
    controlflow: control_flow_1.controlflow,
    dataAccess: data_access_1.dataAccess,
    expressionList: statistics_expression_list_1.expressionList,
    variables: variables_1.variables
};
exports.allFeatureNames = new Set(Object.keys(exports.ALL_FEATURES));
