// @eagleoutice/flowr-dev/r-bridge/shell-executor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RShellExecutor = void 0;
const shell_1 = require("./shell");
const objects_1 = require("../util/objects");
const child_process_1 = require("child_process");
const preload_1 = __importDefault(require("semver/preload"));
const log_1 = require("../util/log");
const init_1 = require("./init");
const convert_values_1 = require("./lang-4.x/convert-values");
const executorLog = log_1.log.getSubLogger({ name: 'RShellExecutor' });
class RShellExecutor {
    options;
    prerequisites;
    constructor(options) {
        this.options = (0, objects_1.deepMergeObject)(shell_1.DEFAULT_R_SHELL_OPTIONS, options);
        this.prerequisites = [(0, init_1.initCommand)(this.options.eol)];
    }
    addPrerequisites(commands) {
        this.prerequisites.push(...(typeof commands == 'string' ? [commands] : commands));
        return this;
    }
    usedRVersion() {
        const version = this.run(`cat(paste0(R.version$major,".",R.version$minor), ${(0, convert_values_1.ts2r)(this.options.eol)})`);
        (0, log_1.expensiveTrace)(executorLog, () => `raw version: ${JSON.stringify(version)}`);
        return preload_1.default.coerce(version);
    }
    run(command, returnErr = false) {
        command += ';base::quit()';
        (0, log_1.expensiveTrace)(executorLog, () => `> ${JSON.stringify(command)}`);
        const returns = (0, child_process_1.spawnSync)(this.options.pathToRExecutable, this.options.commandLineOptions, {
            env: this.options.env,
            cwd: this.options.cwd,
            windowsHide: true,
            encoding: 'utf8',
            input: [...this.prerequisites, command].join(this.options.eol)
        });
        return (returnErr ? returns.stderr : returns.stdout).trim();
    }
}
exports.RShellExecutor = RShellExecutor;
