// @eagleoutice/flowr-dev/cli/repl/server/messages/message-query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestQueryMessage = void 0;
const joi_1 = __importDefault(require("joi"));
const query_schema_1 = require("../../../../queries/query-schema");
exports.requestQueryMessage = {
    type: 'request-query',
    schema: joi_1.default.object({
        type: joi_1.default.string().valid('request-query').required(),
        id: joi_1.default.string().optional().description('If you give the id, the response will be sent to the client with the same id.'),
        filetoken: joi_1.default.string().required().description('The filetoken of the file/data retrieved from the analysis request.'),
        query: query_schema_1.QueriesSchema.required().description('The query to run on the file analysis information.')
    }).description('Request a query to be run on the file analysis information.')
};
