// @eagleoutice/flowr-dev/dataflow/graph/vertex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VertexType = void 0;
exports.isValueVertex = isValueVertex;
exports.isUseVertex = isUseVertex;
exports.isFunctionCallVertex = isFunctionCallVertex;
exports.isVariableDefinitionVertex = isVariableDefinitionVertex;
exports.isFunctionDefinitionVertex = isFunctionDefinitionVertex;
var VertexType;
(function (VertexType) {
    VertexType["Value"] = "value";
    VertexType["Use"] = "use";
    VertexType["FunctionCall"] = "function-call";
    VertexType["VariableDefinition"] = "variable-definition";
    VertexType["FunctionDefinition"] = "function-definition";
})(VertexType || (exports.VertexType = VertexType = {}));
function isValueVertex(vertex) {
    return vertex.tag === VertexType.Value;
}
function isUseVertex(vertex) {
    return vertex.tag === VertexType.Use;
}
function isFunctionCallVertex(vertex) {
    return vertex.tag === VertexType.FunctionCall;
}
function isVariableDefinitionVertex(vertex) {
    return vertex.tag === VertexType.VariableDefinition;
}
function isFunctionDefinitionVertex(vertex) {
    return vertex.tag === VertexType.FunctionDefinition;
}
